src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// On success callback.
+    ///
+    /// This job will be queued by the Faktory server provided
+    /// all the jobs belonging to this batch have been executed successfully.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) success: Option<Job>,
+
+    /// On complete callback.
+    ///
+    /// This job will be queue by the Faktory server after all the jobs
+    /// belonging to this batch have been executed, even if one/some/all
+    /// of the workers have failed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder::new(description)
+    }
+}
+
+impl BatchBuilder {
+    fn build(&mut self) -> Batch {
+        self.try_build()
+            .expect("All required fields have been set.")
+    }
+
+    /// Create a new `BatchBuilder` with optional description of the batch.
+    pub fn new(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder {
+            description: Some(description.into()),
+            ..Self::create_empty()
+        }
+    }
+
+    fn success(&mut self, success_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.success = Some(success_cb.into());
+        self
+    }
+
+    fn complete(&mut self, complete_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.complete = Some(complete_cb.into());
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(&mut self, success_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with only `complete` callback specified.
+    pub fn with_complete_callback(&mut self, complete_cb: Job) -> Batch {
+        self.complete(complete_cb);
+        self.success(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(&mut self, success_cb: Job, complete_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(complete_cb);
+        self.build()
+    }
+}
+
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
+        self.bid.as_ref()
+    }
+
+    pub(crate) fn new(bid: String, prod: &mut Producer<S>) -> BatchHandle<'_, S> {
+        BatchHandle { bid, prod }
+    }
+
+    /// Add the given job to the batch.
+    pub fn add(&mut self, mut job: Job) -> Result<(), Error> {
+        job.custom.insert("bid".into(), self.bid.clone().into());
```

Comment:
Should we error if `job.custom` already contains `bid` for whatever reason?

Reply:
I can imagine this situation, where they decided they need (or they think they are _expected_) to somehow attribute the job to this batch and so put a `bid` into the custom hash, but say a wrong one. My understanding - we should just silently overwrite it and continue.

Reply:
How about making the `Ok` type be `Option<Bid>`, so that we can return the old `Bid` if there was one? The user can always choose to ignore the old value if they don't care.

Reply:
Which is effectively sticking to the [`HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert) API. I see your point!

Reply:
Also added a check to `test_batch_of_jobs_can_be_initiated` test:
```rust
assert!(b.add(job_1).unwrap().is_none());
assert!(b.add(job_2).unwrap().is_none());
assert_eq!(b.add(job_3).unwrap().unwrap(), serde_json::Value("check-check"));
```
where "check-check" is the value of the bid added to `job_3` prior to adding this job to the batch.

Of notice here, is that we are returning an option of serde_json::Value since this is how we keep data in the Job's custom hash. Interestingly, the user is free to add a key `bid` and whatever (implementing into Value of course) value via `add_to_custom_hash`. If they do so, we will return them back their value.

Thinking it over, I can imagine a situation their some domain logic will have a  `bid` which does not stand for batch ID, rather something else - whatever. This is a collision. So, returning a Value that was stored in the hash under the `bid` key will _at least_ attract some attention to this. Also adding to the method's doc of course.